src/applications/SpecificationReader.java

```diff
@@ -30,11 +30,15 @@ private void readChangeOverTimes() {
 
     private void readJobSpecifications() {
```

Comment:
I have been going back and forth on this for a while. There is no two ways about it, `readJobSpecification` in its past state, was a burning dumpster fire: It did so many different things, and it did all in one function. It was very easy to  get overwhelmed by all the logic and for loops. I really appreciate how you broke it up into several pieces, it definitely makes the athstetic of the code blocks better, but I think there is still some big areas for improvement before it gets into master. Here are my concerns:  
* *Function Names:*  Naming functions and variables is hard, and for that reason, you should think twice about creating new functions. If it can't have a short concise and meaningful name, the logic cannot be generalized and that means the code needs some heavy work before it can be its own function.  
* *Side Effects:* Almost all the functions in this class suffer from this. Moreover, this hurts the generality of the functions that you have created and makes these functions seem less dangerous than they are. For example, When I see `readTasks()`, I assume that it is going to read tasks and give them back to me, instead, it reads tasks, puts them in an array, and puts them in a class on the other side of the world, and it does it with a bunch of pointers. In reality, that functionality probably should happen in `readJobSpecification` (which is due for a new name), right where the person who is reading that function can see its side effecting nature.  

In summary, I  would like to see a few things with this: More thought about the generality of these functions, less side effects in the bowls of helper functions, and more honest variable names.